### PR TITLE
Add multi-theme support to Polaris

### PR DESCRIPTION
**Pull request by aaronccasanova**
This PR consolidates the integration of three feature branches that collectively introduce a new base/variant theme architecture to Shopify Polaris. The merged PRs introduce multi-theme support, updated build artifacts, and a new `useTheme` hook for runtime Polaris tokens access. Below, you will find a brief overview of each PR, with links to the original PRs for a more detailed review.

## [Add multi-theme support to Polaris tokens](https://github.com/Shopify/polaris/pull/10103)

This PR introduces multi-theme support to `@shopify/polaris-tokens` by implementing a base/variant architecture. The base theme serves as the foundation for multiple variant themes, which extend the base theme and are the only themes exposed to consumers. Simplifying system changes by flattening the relationship between themes is the primary advantage of this pattern. The PR also includes implementation details like directory structure, stylesheet generation, and variant theme functionality.

**Example `styles.scss` updates**

```scss
:root {
  /* Default variant theme custom properties (light) */
}

html.Polaris-Summer-Editions-2023 {
  /* Variant theme custom properties */
  /*
     Note: The above selector is special
     cased for backward compatibility. See
     below for updated theme selector structure.
  */
}

html.p-theme-light-high-contrast {
  /* Variant theme custom properties */
  /*
     Note: Variant theme selectors
     contain a subset of custom properties
     overriding the base theme
  */
}

html.p-theme-dark {/* ... */}
html.p-theme-dark-high-contrast {/* ... */}
html.p-theme-dim {/* ... */}
```

## [Add multi-theme build artifacts to Polaris tokens](https://github.com/Shopify/polaris/pull/10250)

This PR is a rework of #10153, redefining how the `themes` are accessed from `@shopify/polaris-tokens`. It introduces changes on how asset builder functions are written and replaces reference to `themes.light` with a `themeDefault` alias.

```ts
import {themes} from '@shopify/polaris-tokens';

themes['Polaris-Summer-Editions-2023'].color['color-bg']
```

> Note: Future releases can reduce the bundle size impact of including all themes on one object and enable more flexibility (for example with a `createTheme` util that deep merges variant theme partials). However, that is not needed at this time and can be introduced in a minor release.

## [Add `useTheme` hook for runtime Polaris tokens access](https://github.com/Shopify/polaris/pull/10252)

This PR is a rework of #10229, brings updates to the `AppProvider` by incorporating a new `ThemeContext.Provider` and an associated `useTheme` hook.

```tsx
import {useTheme} from '@shopify/polaris';

function App() {
  const theme = useTheme()

  theme.color['color-bg'] // '#123'
}
```

### CODE REVIEW

File: polaris-tokens/package.json [hunk 1]
```diff
@@ -55,5 +55,8 @@
   ],
   "files": [
     "dist"
-  ]
+  ],
+  "dependencies": {
+    "deepmerge": "^4.3.1"
```
Comment (aaronccasanova) [inline polaris-tokens/package.json:60]: This is needed for backward compatibility with the existing `metadata` export and can be moved to a `devDependency` post v12.

File: polaris-tokens/scripts/toValues.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+  const themeDefault = extractMetaThemeValues(metaThemeDefault);
+
+  await fs.promises.writeFile(
+    path.join(outputDir, 'index.ts'),
+
+    [
+      `export * from '../src/index';`,
+      Object.entries(themeDefault).map(createExport),
+      createExport(['tokens', themeDefault]),
```
Comment (aaronccasanova) [inline polaris-tokens/scripts/toValues.ts:23]: Note: These exports will be removed in the v12 release as they will no longer be relevant in the new theming architecture.

**Review by lgriffee — APPROVED** [review]

**Pull request merged**